File: src/main/java/jv/com/tencent/cain/design_pattern/observer/Subject.java
package jv.com.tencent.cain.design_pattern.observer;

/**
 * @author : jiangyu
 * @date : 2020/3/3
 * @desc : xxx
 */
public interface Subject {
    void addObserver(Observer observer);

    void removeObserver(Observer observer);

    void removeAll();

    void notifyAllObserver(Object data);

    void notify(Observer observer, Object data);
}

File: src/main/kotlin/kt/com/tencent/cain/desigin_pattern/observer/Subject.kt
package kt.com.tencent.cain.desigin_pattern.observer

/**
 * @author : jiangyu
 * @date   : 2020/3/3
 * @desc   : xxx
 */
interface Subject {
    fun addObserver(observer: Observer)

    fun remove(observer: Observer)

    fun removeAll()

    fun notifyAll(data: Any)

    fun notify(observer: Observer, data: Any)
}